### ISSUE TITLE
BUG: incorrect reporting of fcTL::delay_num

### ISSUE THREAD

**Issue by jbowler** [svgeesus/pngcheck#14]
In Philip's 001.png I see the following values in fcTL[0]:

- width 128
- height 64
- x_offset 0
- y_offset 0
- dispose 0
- blend 1
- delay_num 100
- delay_den 100

So the delay is, from my code, 1s, however your code reports delay_num as 6553700.  I checked the file and it does seem to me that both U16's are "100"; delay_den is not 0 (which would result in the same value IRC), so I'm guessing it is your bug not mine :-)

**Comment by jbowler** [comment]
Ok, I'm not going to raise any more separate issues:

007.png: incorrectly reported as invalid - pngcheck gives out on the first frame after IDAT (it's valid because my program reports a following fdAT then fcTL and a final fdAT).

This seems to continue with the later test images; none of them get beyond the first fcTL after IDAT if there is an fcTL **before** the first IDAT.

In my version of pngcheck there was a very useful **-f** flag which continued past errors; so it would continue past a detected error in a chunk and keep printing the rest.

It would be nice if the author would, in fact, put 042 on the web site; I can see disabling it but for testing purposes crashers are the most useful.  STO does not work.

**Comment by svgeesus** [comment]
> So the delay is, from my code, 1s, however your code reports delay_num as 6553700. I checked the file and it does seem to me that both U16's are "100"; delay_den is not 0 (which would result in the same value IRC), so I'm guessing it is your bug not mine :-)

Agreed, that is a bug.  It reports 

> Frame delay 65537 (6553700 / 100) sec

Aha! I see the problem:

```C
if (no_err(kMinorError)) {
        sequence_number = LG(buffer);
        frame_width = LG(buffer+4);
        frame_height = LG(buffer+8);
        x_offset = LG(buffer+12);
        y_offset = LG(buffer+16);
        delay_num = LG(buffer+20); 
        delay_den = SH(buffer+22);
        dispose_op = *(buffer+24);
        blend_op = *(buffer+25);
      }
```

Should be `SH` not `LG`

**Comment by svgeesus** [comment]
> 007.png: incorrectly reported as invalid - pngcheck gives out on the first frame after IDAT (it's valid because my program reports a following fdAT then fcTL and a final fdAT).

I'm not getting an error:

```
chris@SuperNomad:/mnt/c/Users/chris/Documents/PNG/philip_APNG_tests_files$ pngcheck -c -v 007.png
File: 007.png (617 bytes)
  chunk IHDR at offset 0x0000c, length 13
    128 x 64 image, 32-bit RGB+alpha, non-interlaced
  chunk acTL at offset 0x00025, length 8
    Animated PNG, 3 frames, plays 1 times
  chunk fcTL at offset 0x00039, length 26
    Frame, sequence number 0
    Width 128, height 64 starting at (0, 0)
    Frame delay 6554 (655460 / 100) sec
    No disposal before next frame
    Frame composites (source over) with buffer
  chunk IDAT at offset 0x0005f, length 147
    zlib: deflated, 32K window, default compression
  chunk fcTL at offset 0x000fe, length 26
    Frame, sequence number 1
    Width 128, height 64 starting at (0, 0)
    Frame delay 6554 (655460 / 100) sec
    No disposal before next frame
    Frame composites (source over) with buffer
  chunk fdAT at offset 0x00124, length 196
    Frame data, sequence number 2
  chunk fcTL at offset 0x001f4, length 26
    Frame, sequence number 3
    Width 128, height 64 starting at (0, 0)
    Frame delay 6554 (655460 / 100) sec
    No disposal before next frame
    Frame composites (source over) with buffer
  chunk fdAT at offset 0x0021a, length 59
    Frame data, sequence number 4
  chunk IEND at offset 0x00261, length 0
No errors detected in 007.png (9 chunks, 98.1% compression).
```

**Comment by jbowler** [comment]
Hum.  Most odd:

`jbowler@hippopopus ~/src/libapng/git/libapng $ ./test-core-read <tests/*/007.png
INFO: this is an APNG with 3 frames x 1 loops
INFO: fcTL[0] BEFORE IDAT
INFO: fcTL[0(0)]: (128x64)+(0,0) DISPOSE(none(0)) BLEND(Porter-Duff(over)(1)) DELAY 0.1s
INFO: fcTL[1(1)]: (128x64)+(0,0) DISPOSE(none(0)) BLEND(Porter-Duff(over)(1)) DELAY 0.1s
INFO: fdAT[2] data(0x558cb2be1354)[192] fcTL[1(1)]
INFO: fcTL[2(3)]: (128x64)+(0,0) DISPOSE(none(0)) BLEND(Porter-Duff(over)(1)) DELAY 0.1s
INFO: fdAT[4] data(0x558cb2be1524)[55] fcTL[2(3)]
INFO: APNG end fcTL+fdAT[5 chunks]: 3 frames; last: 3
jbowler@hippopopus ~/src/libapng/git/libapng $ pngcheck -v tests/*/007.png
File: tests/philip.html5.org/007.png (617 bytes)
  chunk IHDR at offset 0x0000c, length 13
    128 x 64 image, 32-bit RGB+alpha, non-interlaced
  chunk acTL at offset 0x00025, length 8
    Animated PNG, 3 frames, plays 1 times
  chunk fcTL at offset 0x00039, length 26
    Frame, sequence number 0
    Width 128, height 64 starting at (0, 0)
    Frame delay 6554 (655460 / 100) sec
    No disposal before next frame
    Frame composites (source over) with buffer
  chunk IDAT at offset 0x0005f, length 147
    zlib: deflated, 32K window, default compression
  chunk fcTL at offset 0x000fe, length 26:  At least one fdAT is required for each frame (except first, if IDAT)
ERRORS DETECTED in tests/philip.html5.org/007.png`

I made changes to my program but not to Mr Bond.

**Comment by jbowler** [comment]
You fixed it with this check-in:

8d990187d5060ad79a297073ef42dab80a180afe

**Comment by jbowler** [comment]
Closed as fixed.
